### PR TITLE
Enable cherrypick plugin for k-sigs/{cluster-api,controller-tools,controller-runtime}

### PR DESCRIPTION
**Pull request by droslean**
Enable cherrypick plugin for k-sigs/{cluster-api,controller-tools,controller-runtime}

### CODE REVIEW

**Comment by alvaroaleman** [comment]
/uncc

I do not have a say in prow.k8s.io config settings :upside_down_face:

**Comment by justaugustus** [comment]
Explicit hold:
/hold

What's the context/justification and was this discussed with the @kubernetes/patch-release-team?

**Comment by droslean** [comment]
We found ourselves the need to use the cherrypick plugin in https://github.com/kubernetes/kubernetes/pull/88099

Also, there wasn't any discussion with @kubernetes/patch-release-team.  We discussed this in the last weekly release meeting but I thought that the plugin was already enabled. 

@justaugustus How we should proceed?

**Comment by fejta** [comment]
Technically looks fine, but can we get some approval from `sig-contribex` that this is wanted across the org? (Seems useful to me)

**Comment by cheftako** [comment]
> We found ourselves the need to use the cherrypick plugin in [kubernetes/kubernetes#88099](https://github.com/kubernetes/kubernetes/pull/88099)

In the short term I ran the cherry-pick script and created https://github.com/kubernetes/kubernetes/pull/88278 for 1.17

**Comment by spiffxp** [comment]
/cc @justaugustus @cblecker 
I think this would be useful to enable for the org, or at least trialing in kubernetes/kubernetes. 
 It would greatly reduce the friction necessary to cherry-pick a given PR into a release branch

What concerns would you have about enabling this, and what do you think next steps should be?

ref: https://github.com/kubernetes/test-infra/tree/master/prow/external-plugins/cherrypicker

File: config/prow/plugins.yaml
Comment (justaugustus) [inline config/prow/plugins.yaml]: Is this the right URL?

**Comment by spiffxp** [comment]
>Did we have a cherrypick plugin before? I feel like we did and turned it off? Is this net new or turning back on the one we removed?

This is net new for kubernetes, though it's been in use for openshift and istio for a while.

I think you're remembering a mungegithub plugin that auto-applied the cherry-pick-approved label when hard-to-understand preconditions were met. We turned that off and never migrated it to prow, prior to migrating kubernetes/kubernetes to tide.

>What are the failure modes of the plugin?

You might need to be more specific here.  It won't open a PR if one already exists. It will comment on the source PR if a cherrypick failed, and there's discussion about having it file an issue (https://github.com/kubernetes/test-infra/issues/16487)

>What's the guidance for using this plugin? Is it documented in an easy-to-find place for contributors?

I would suggest we trial to ensure we catch whatever other failure modes you're concerned about, and then update documentation to make this the recommended approach as a followup. If enabled, it would show up in http://go.k8s.io/bot-commands (eg: https://prow.svc.ci.openshift.org/command-help#cherrypick)

>How do I edit a CP PR that's not quite a mirror of the parent PR if a bot is creating them?

Here I would recommend closing the PR and opening your own as we do now. The most GitHub could do for us is allow people with repo write access to add new commits to the branch.

>Who creates/manages these cherrypick/XXX labels (that currently don't exist)?

My suggestion would be that we trial using this without labels, and solely as a `/cherrypick` command

/cc @stevekuznetsov @clarketm 
is there any additional context you could give from openshift or istio's usage of this plugin?

/hold
FYI @droslean @fejta the cherrypicker service isn't deployed to prow.k8s.io, we would need something equivalent to these files added to prow/cluster/, which should probably done in a separate PR?
- https://github.com/istio/test-infra/blob/7807fbc912b2873050b51aeb4b72235aa706df19/prow/cluster/cherrypicker_deployment.yaml
- https://github.com/istio/test-infra/blob/974ccb4ebe99c8b724e257eaeeec634526c4a531/prow/cluster/cherrypicker_service.yaml

**Comment by droslean** [comment]
@BenTheElder @spiffxp Deployment and service added. Can you check?

**Comment by cblecker** [comment]
Acknowledging that I’ve seen this and will review this week.

**Comment by BenTheElder** [comment]
Not sure why I'm on this PR, there are already 5 people assigned and this isn't my decision ... 
Cherry pick automation sounds potentially useful though 🤷‍♂ 
/uncc
/assign @cblecker

**Comment by justaugustus** [comment]
Thanks for the additional context, @spiffxp!
Approved on the Patch Release Team side.

**Comment by clarketm** [comment]
> is there any additional context you could give from openshift or istio's usage of this plugin?

One more thing to add. On occasion, a cherrypick is silently missed due to a `git` _not our ref_ error: https://github.com/istio/test-infra/issues/2309. This error occurs much less frequently when using the `/cherrypick` command than the labels; and more frequently in larger repos – ones that change more frequently and have a higher quantity of refs (e.g. istio/istio). I created https://github.com/kubernetes/test-infra/issues/16130 but it is still a wip.

**Comment by justaugustus** [comment]
Thanks for the heads up on that one, @clarketm!

**Comment by justaugustus** [comment]
@cblecker @BenTheElder @spiffxp -- Ping on this one

**Comment by cblecker** [comment]
This is still on my radar, I've just had some more pressing fires to put out.

**Comment by justaugustus** [comment]
SG, thanks!

**Comment by BenTheElder** [comment]
@justaugustus https://github.com/kubernetes/test-infra/pull/16329#issuecomment-599255263

**Comment by cblecker** [comment]
Sorry for the delay in getting back to this one.

My biggest concern with this plugin has been that it doesn't work 100% of the time (you might need to pick multiple PRs together, or there may be rebase issues, etc), and then you need two separate ways to cherry pick (the script can't go away completely). This might make like.. 50-60% of picks quicker (total guess), but there will still be ones you need to fall back to some other method.

What I think we should possibly look at is a trial.. X number of months/releases, and then we look at the data.. how many picks were successful? How many folks used this method, as opposed to the script? We would also need to update the documentation.

So.. who is going to drive this? sig-testing? sig-release? I don't have the bandwidth to drive directly, but I'm happy to help define parameters so we know if this is improving/detracting from the contributor experience/workflow.

cc @nikhita if she has any thoughts

**Comment by LappleApple** [comment]
Heya @nikhita, pinging on this wrt @cblecker's comment ^^ -- WDYT?

As for whether sig-release takes this on or not, cc'ing  @tpepper @saschagrunert @alejandrox1 @justaugustus to ask whether we might include this in our sig-release pruning session next week — either as an item or as something to keep in mind while we prune.

**Comment by alvaroaleman** [comment]
@justaugustus @cblecker any chance we can activate this at least for k-sigs? I don't really see much of a drawback of having this, if it doesn't work you can still manually cherrpick, if it does work it is helpful

**Comment by saschagrunert** [comment]
/assign
@justaugustus do you think we can merge this one?

**Comment by spiffxp** [comment]
/test pull-test-infra-integration
#20498 (comment)

**Comment by spiffxp** [comment]
@cblecker 

>What I think we should possibly look at is a trial.. X number of months/releases, and then we look at the data.. how many picks were successful? How many folks used this method, as opposed to the script? We would also need to update the documentation.

>So.. who is going to drive this? sig-testing? sig-release? I don't have the bandwidth to drive directly, but I'm happy to help define parameters so we know if this is improving/detracting from the contributor experience/workflow.

How about we see if there are any subprojects interested in piloting this before rolling out to broader org?
- @alvaroaleman did you have any specific k-sigs repos in mind? 
- @saschagrunert any release related repos in mind?

**Comment by alvaroaleman** [comment]
> @alvaroaleman did you have any specific k-sigs repos in mind?

Yes, would love this for https://github.com/kubernetes-sigs/controller-runtime, cc @vincepri

**Comment by vincepri** [comment]
I'd add, https://github.com/kubernetes-sigs/cluster-api and https://github.com/kubernetes-sigs/controller-tools

**Comment by saschagrunert** [comment]
> * @saschagrunert any release related repos in mind?

Sorry for the late reply. We do not use cherry-picking within any release related repository right now, so I don't think we can test the plugin there in a feasible way, yet.

**Comment by alvaroaleman** [comment]
@spiffxp @cblecker what is blocking this from getting enabled in k sigs or a k/sigs repos that want it? Having this would be tremendously helpful in maintaing controller-runtime and it appears its blocked by the question if its useful for k/k which IMHO is entirely unrelated?

**Comment by cblecker** [comment]
This was initially about org-wide plug-in changes. I don't think I have any outstanding concerns on for specific k-sigs/ repos that dont follow the k/k cherry picking plan/cadence.

@spiffxp @nikhita Do either of you?

**Comment by spiffxp** [comment]
No blocker. I thought the pr hasn't been updated for those specific repos. I will happily approve something specific

File: config/prow/cluster/cherrypicker_deployment.yaml
Comment (cblecker) [inline config/prow/cluster/cherrypicker_deployment.yaml:65]: Is this the right token?

Comment (alvaroaleman) [reply config/prow/cluster/cherrypicker_deployment.yaml:65]: I think so: https://github.com/kubernetes/test-infra/blob/2ee80aecbafb008d28756c0399b582ac20dc2775/config/prow/cluster/needs-rebase_deployment.yaml#L59-L61

**Comment by alvaroaleman** [comment]
/retitle Enable cherrypick plugin for k-sigs/{cluster-api,controller-tools,controller-runtime}

**Review by alvaroaleman — APPROVED** [review]
/lgtm
for controller-runtime

File: config/prow/cluster/cherrypicker_deployment.yaml
Comment (cblecker) [reply config/prow/cluster/cherrypicker_deployment.yaml:65]: I mean, is k8s-ci-robot the right account for us to be doing this under

Comment (alvaroaleman) [reply config/prow/cluster/cherrypicker_deployment.yaml:65]: True, good point. Is there one that can be used?

Comment (alvaroaleman) [reply config/prow/cluster/cherrypicker_deployment.yaml:65]: I asked about this in the latest test-infra and @spiffxp mentioned that https://github.com/k8s-bot can be used. But I don't see any reference to a token of that bot in the repo, am I missing it?  @chaodaiG can you check if you see sth on the cluster?

File: config/prow/plugins.yaml
Comment (hasheddan) [inline config/prow/plugins.yaml:1084]: @saschagrunert @justaugustus I think we can remove this from release engineering concern right now given that currently none of these repos where the plugin would be enabled our under or purview. What do you think?

File: config/prow/cluster/cherrypicker_deployment.yaml
Comment (chaodaiG) [reply config/prow/cluster/cherrypicker_deployment.yaml:65]: There is only one `oauth-token` like secret in k8s prow service cluster, and I believe that belongs to `k8s-ci-robot`

File: config/prow/plugins.yaml
Comment (saschagrunert) [reply config/prow/plugins.yaml:1084]: Agree :+1:

**Comment by vincepri** [comment]
Hi folks, what can we do to get this PR merged? Are there any outstanding tasks? @spiffxp @justaugustus 

cc @dims

**Comment by vincepri** [comment]
/lgtm

for CR and Cluster API

File: config/prow/cluster/cherrypicker_deployment.yaml
Comment (vincepri) [inline config/prow/cluster/cherrypicker_deployment.yaml:1]:
```suggestion
# Copyright 2021 The Kubernetes Authors All rights reserved.
```

File: config/prow/cluster/cherrypicker_service.yaml
Comment (vincepri) [inline config/prow/cluster/cherrypicker_service.yaml:1]:
```suggestion
# Copyright 2021 The Kubernetes Authors All rights reserved.
```

**Comment by dims** [comment]
looks like we have sig-release signoff ( https://github.com/kubernetes/test-infra/pull/16329/files#r612485076 )

File: config/prow/cluster/cherrypicker_deployment.yaml
Comment (dims) [reply config/prow/cluster/cherrypicker_deployment.yaml:65]: Looks like this is the right token then!

File: config/prow/plugins.yaml
Comment (dims) [reply config/prow/plugins.yaml:1084]: thanks @hasheddan @saschagrunert

File: config/prow/cluster/cherrypicker_deployment.yaml
Comment (alvaroaleman) [reply config/prow/cluster/cherrypicker_deployment.yaml:65]: No it is unfortunately not, we can not use the same account as Prow itself as Prow ignores events from itself.

**Comment by dims** [comment]
/hold cancel
/approve
/lgtm

we can iterate if needed (token?)

**Comment by dims** [comment]
/assign @cblecker @spiffxp 

Christoph, Aaron,
ok with this? it's been struggling for some time.

**Pull request merged**
